### PR TITLE
[batch] fix n cancelled creating jobs

### PR DESCRIPTION
**Pull request by danking**
I ran the select ... from (select ...) part of the UPDATE which fixes
this issue. It completed in ~72 seconds, so I think this change is actually
plenty fast to run.

### CODE REVIEW

**Comment by jigold** [comment]
Can you also fix `check_incremental` in batch/batch/driver/main.py to actually throw an exception if there's a problem?

File: batch/sql/fix-n-cancelled-creating-jobs.sql [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+       t.n
+FROM (
+  select batches.user, jobs.inst_coll, count(*) as n
+  from jobs
+  left join batches_cancelled
+  on jobs.batch_id = batches_cancelled.id
+  left join batches
+  on jobs.batch_id = batches.id
+  where jobs.state = 'Creating'
+    and (NOT jobs.always_run AND (jobs.cancelled OR (batches_cancelled.id is not null)))
```
Comment (jigold) [inline batch/sql/fix-n-cancelled-creating-jobs.sql:211]: caps lock please

**Comment by danking** [comment]
Yep! https://github.com/hail-is/hail/pull/11279

File: batch/sql/fix-n-cancelled-creating-jobs.sql [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        n_creating_cancellable_jobs = n_creating_cancellable_jobs - 1;
+    END IF;
+
+    # state = 'Creating' jobs cannot be cancelled at the job level
+    IF NOT OLD.always_run AND cur_batch_cancelled THEN
+      # cancelled
+      INSERT INTO user_inst_coll_resources (user, inst_coll, token, n_cancelled_creating_jobs)
+      VALUES (cur_user, OLD.inst_coll, rand_token, -1)
+      ON DUPLICATE KEY UPDATE
+        n_cancelled_creating_jobs = n_cancelled_creating_jobs - 1;
```
Comment (danking) [inline batch/sql/fix-n-cancelled-creating-jobs.sql:86]: this is the important line.

File: batch/sql/fix-n-cancelled-creating-jobs.sql (hunk 1 above)
Comment (jigold) [inline batch/sql/fix-n-cancelled-creating-jobs.sql:211]: I think this is right.

**Review by jigold — CHANGES_REQUESTED** [review]

File: batch/sql/fix-n-cancelled-creating-jobs.sql [hunk 3]
```diff
@@ -201,15 +201,15 @@ select t.user,
...
-  select batches.user, jobs.inst_coll, count(*) as n
-  from jobs
-  left join batches_cancelled
-  on jobs.batch_id = batches_cancelled.id
-  left join batches
-  on jobs.batch_id = batches.id
-  where jobs.state = 'Creating'
-    and (NOT jobs.always_run AND (jobs.cancelled OR (batches_cancelled.id is not null)))
-  group by batches.user, jobs.inst_coll
+  SELECT batches.user, jobs.inst_coll, count(*) AS n
```
Comment (jigold) [inline batch/sql/fix-n-cancelled-creating-jobs.sql:204]: So this only updates the row with token = 0. What happens to the other rows? I think we need to set all other rows to 0.

**Review by jigold — CHANGES_REQUESTED** [review]

File: batch/sql/fix-n-cancelled-creating-jobs.sql (hunk 3 above)
Comment (danking) [reply batch/sql/fix-n-cancelled-creating-jobs.sql:204]: I do, the pervious statement in this transaction is an UPDATE that sets them all to 0.

File: batch/sql/fix-n-cancelled-creating-jobs.sql [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  `n_ready_jobs`,
+  `n_running_jobs`,
+  `n_creating_jobs`,
+  `ready_cores_mcpu`,
+  `running_cores_mcpu`,
+  `n_cancelled_ready_jobs`,
+  `n_cancelled_running_jobs`,
+  `n_cancelled_creating_jobs`
+)
+select t.user,
```
Comment (jigold) [inline batch/sql/fix-n-cancelled-creating-jobs.sql:192]: One more

**Review by jigold — CHANGES_REQUESTED** [review]

**Comment by jigold** [comment]
I have to go, but what happens when you update the rows to 0. Does the trigger run? I presume the n_cancelled_creating_jobs stay at 0 because now the additions / subtractions are in sync?

**Comment by danking** [comment]
@jigold there are no user_inst_coll_resources table triggers in the database. Are you referring to something other than a db trigger?

**Review by jigold — APPROVED** [review]

**Comment by jigold** [comment]
I was getting confused where the trigger was located.

**Comment by jigold** [comment]
I put the WIP back on -- it's too late in the day to react if there's an issue and the database is in a bad state.

**Pull request merged**